pkg/controlplane/peer_endpoint_controller.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlplane
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	utilpeerproxy "k8s.io/apiserver/pkg/util/peerproxy"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controlplane/reconcilers"
+	corerest "k8s.io/kubernetes/pkg/registry/core/rest"
+	"k8s.io/kubernetes/pkg/util/async"
+)
+
+// PeerEndpointController is the controller manager for updating the Kubernetes service
+// endpoints. This is essentially a clone of the controller manager for the core
+// bootstrap Kubernetes controller loops.
+
+// We have cloned the controller to have a separate independent endpoint reconciliation loop
+// which ensures that the peer kube-apiservers are fetching the updated endpoint info for a given apiserver
+// in the case when the peer wants to proxy the request to the given apiserver because it can not serve the
+// request itself due to version mismatch.
+type PeerEndpointController struct {
+	client                 kubernetes.Interface
+	PeerEndpointReconciler reconcilers.PeerEndpointReconciler
+	EndpointInterval       time.Duration
+	PublicIP               net.IP
+	ServerId               string
+	PublicServicePort      int
+	// PeerBindAddress stores the IP and port of this kube-apiserver. Used by peer kube-apiservers to
+	// route request to this apiserver in case of a version skew.
+	PeerBindAddress utilpeerproxy.PeerBindAddress
+	runner          *async.Runner
+}
+
+// NewPeerEndpointController returns a controller for watching the core capabilities of the server
+func (c *completedConfig) NewPeerEndpointController(legacyRESTStorage corerest.LegacyRESTStorage, client kubernetes.Interface) (*PeerEndpointController, error) {
+	_, publicServicePort, err := c.GenericConfig.SecureServing.HostPort()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get listener address: %w", err)
+	}
+
+	return &PeerEndpointController{
+		client:                 client,
+		PeerEndpointReconciler: c.GenericConfig.PeerEndpointReconciler,
+		EndpointInterval:       c.ExtraConfig.EndpointReconcilerConfig.Interval,
+		PublicIP:               c.GenericConfig.PublicAddress,
```

Comment:
I thought this could be overridden via the  peerBindAddress.PeerBindIP

Reply:
oh, I see, you pass all the config and discriminate inside the update method

Reply:
yeah keeping the publicIP and peerbindaddress separate and not overwriting one with another. publicIP is stored as the [corev1.EndpointAddress.IP](https://github.com/kubernetes/kubernetes/blob/8ffbbe455b30d030a8c99857ad85282f6aec95e6/pkg/apis/core/types.go#L4324)  and peerbind address is saved as an annotation.

Reply:
refactored to only use a single address (either public IP, port or peer-advertise-address passed in flags)